Update listing details to accommodate AMI tiers  #171215384

akegan opened the pull request:
https://www.pivotaltracker.com/story/show/171215384

akegan commented on app/assets/javascripts/listings/ListingUnitService.js.coffee:
```diff
@@ -29,49 +34,126 @@ ListingUnitService = ($http, ListingConstantsService, ListingIdentityService) ->
-  #     {'1br': [{rent/price, # of units, etc}, ...], '2br': [...]},
-  #   '40 (percent ami)': {...}}
+  Service._getAMIAmount = (incomeList, occupancy) ->
+    if incomeList && incomeList.values?
```
Not sure how best to handle cases where we can't find the AMI chart associated with the unit. This happens fairly rarely. Console.error should help it show up in Sentry, but I'm wondering if we should do something a bit more drastic so that it looks wrong for SMEs when they're previewing the listing.

james-wills-sf replied:
maybe we can check in on the sentry logs in a couple weeks to see if this ever happens? If it has happened we can think about error handling.

Unless you know for a fact that this will happen

akegan commented on app/assets/javascripts/listings/ListingUnitService.js.coffee:
```diff
@@ -29,49 +34,126 @@ ListingUnitService = ($http, ListingConstantsService, ListingIdentityService) ->
+    # occupancy options.
+    maxOccupancy = unitGroup.Max_Occupancy || unitGroup.Min_Occupancy + 2
+    occupancyRange = [unitGroup.Min_Occupancy..maxOccupancy]
+    # FIXME: This does not handle cases where listings have multiple AMI charts with the same percent.
```
right now I think what we do in the max income table is just pick one of the charts if there is > 1 kind of chart with the same year + percent. It's pretty rare (I couldn't find any that are live now). 

Wondering if it makes sense to fix in this PR or do in a follow-up ticket.

james-wills-sf replied:
This PR is pretty big, I'd be happy with splitting it into a separate PR if that's not too much of a hassle

james-wills-sf commented:
@akegan just finished QA and I found the following issues:

### 1. No rent price is listed for % income rents
This is probably the same issue you already found.

**reproducible listing id:** a0W0P00000Gs3V3UAJ
<img src="https://user-images.githubusercontent.com/64036574/83053634-e544e680-a005-11ea-8da6-cfc1ba83d116.png" width="330" />

### 2. Borders don't go all the way across the view on small screen sizes
Only able to reproduce on some of the borders for certain listings
**reproducible listing ID:** a0W1D000001IJ6SUAW
<img src="https://user-images.githubusercontent.com/64036574/83053770-17eedf00-a006-11ea-9961-1cdff285432f.png" width="330" />

### 3. On small screens, not enough spacing between value and labels
Could add a [non-breaking space](https://www.computerhope.com/jargon/n/nbsp.htm) between the spans.
<img src="https://user-images.githubusercontent.com/64036574/83053909-5b494d80-a006-11ea-84d1-80820e749c19.png" width="330" />

akegan commented:
@cpalmatier  for the case of % income I'm also wondering if the minimum income should be $0 or should be something like "None", did you have a design for that?

akegan commented on app/assets/javascripts/config/angularRoutes.js.coffee:
```diff
@@ -123,8 +123,9 @@
               # trigger this asynchronously, allowing the listing page to load first
-              setTimeout(ListingDataService.getListingAMI.bind(null, ListingDataService.listing))
-              setTimeout(ListingUnitService.getListingUnits.bind(null, ListingDataService.listing, forceRecache))
+              ListingUnitService.getListingAMI(ListingDataService.listing).then( ->
```
we need the AMIs to be able to calculate the units now, so I wanted to make sure that amis were present before we started grouping units. Could probably be done in a more sophisticated way in the service instead.

For the timeout, I was honestly just copying what was out there for the listing call. I wasn't sure if we could wrap both in setTimeout.

I'd like to maybe follow up with a ticket that does a better job of optimizing the calculations so there's less of a delay when the page loads

akegan commented:
OK I've fixed the issue with % income

I tried adding `&nbsp;`s everywhere to space out the #s from the "small" labels but realized that this is an issue on the listing directory too so we'd have to update it in ~15 places. If we can update the pattern lib that might be a better long-term solution.

cpalmatier commented:
@akegan answers re: a few of the above comments...

- $0 minimum income requirements: those should just be "None" with no "per month" added.
- the <small> spacing thing: i thought we ran into (and fixed) that before? in the pattern library at the "small" (confusing!) breakpoint, there was a rule added for margin-left: 0.125rem for <small> elements in tables with class .pricing-table. happy to take a look at the review app and diagnose.
- the border thing in @james-wills-sf 's #2 above is something i _think_ is fairly straightforward and would be a fine fast-follow if y'all don't think it's a showstopper.

akegan commented:
Thanks @cpalmatier  I will make the change to None (i think the code is in there but isn't quite working)

I thought we had solved the spacing thing too 🤔. Here is the review app if you're up for checking it out https://dahlia-web-full-pr-1342.herokuapp.com/listings/a0W0P00000Gs3V3UAJ

We can wait + follow-up on the border thing, maybe smoosh it together with the "moderate income" overlap fix.

akegan commented:
OK @james-wills-sf @cpalmatier I think I fixed all the issues we were going to fix in this pr:
1. Show rent if it's % income
1. Show "none" for min income if it's $0 on rental and sales
1. Added spacing on mobile view (thanks to Chris!)

let me know if it looks good and I'll merge!
